### PR TITLE
track allocate/free event sequences in CudaCachingAllocator

### PR DESCRIPTION
**Pull request by o-hanna**
Towards visualizing and solving fragmentation problems, CudaCachingAllocator is modified to track the allocate and free event sequences. An API is exposed to return the traced sequence.

- Use torch.cuda.get_alloc_free_events() to get the traced allocate/free event sequence.
- torch.cuda.get_alloc_free_events() returns a list of allocate/free events of each device.
- Allocate/free events of each device are returned as list of structures with members:
- - ptr: pointer to assigned block
- - size: size of allocation in bytes. The size is positive for allocation events and negative for free events.

Examples:
- torch.cuda.get_alloc_free_events()[0]: list of allocate/free events of device 0
- torch.cuda.get_alloc_free_events()[0][n].size: the size of allocation/free (with negative value to indicate free) for the (n+1)-th event for device 0

The exposed API is tested on pytorch/examples/mnist. The returned allocate/free sequences match the allocate/free events when logged from CudaCachingAllocator.cpp.

### CODE REVIEW

**Review by mrshenli — CHANGES_REQUESTED** [review]
Hey @o-hanna, could you please target this PR to a non-master branch (I.e., create your own branch in PyTorch repo)? 

You can change the target branch by clicking the "Edit" button next to the PR title. 

Screen Shot 2022-07-20 at 12 59 17 PM

**Comment by H-Huang** [comment]
@o-hanna As we discussed, here is how to debug the failures on this PR:

**Debug C++ test:**

image

1. Test is here: https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/codegen/cuda/test/test_gpu.cpp#L22984
2. Run test with `./build/bin/test_jit --gtest_filter=NVFuserMultithreadedTest.SingleFunction_CUDA`

**Debug Python test:**

image

1. Test is here: https://github.com/pytorch/pytorch/blob/master/test/distributed/test_c10d_gloo.py#L1044
2. Run test with `gpurun pytest test/distributed/test_c10d_gloo.py -vs -k test_gather_stress_cuda` (pip install pytest beforehand)

File: c10/cuda/CUDACachingAllocator.h
Comment (mrshenli) [inline c10/cuda/CUDACachingAllocator.h:131]: this file uses snake instead of camel naming

File: c10/cuda/CUDACachingAllocator.cpp
Comment (mrshenli) [inline c10/cuda/CUDACachingAllocator.cpp:500]: why do you need manually resize vector instead of letting vector handle it automatically?

File: c10/cuda/CUDACachingAllocator.cpp
Comment (mrshenli) [inline c10/cuda/CUDACachingAllocator.cpp:510]: this can be a const function?

File: c10/cuda/CUDACachingAllocator.h
Comment (mrshenli) [inline c10/cuda/CUDACachingAllocator.h:131]: for camel naming, it should be `getAllocFreeEvents`

File: c10/cuda/CUDACachingAllocator.cpp
Comment (o-hanna) [reply c10/cuda/CUDACachingAllocator.cpp:500]: I noticed that the number of events can large. If the vector is extended by one element each iteration and the running time becomes quadratic, this can be slow. I double the size when we need more space to maintain a linear amortized running time.

File: c10/cuda/CUDACachingAllocator.h
Comment (o-hanna) [reply c10/cuda/CUDACachingAllocator.h]: Done. Thank you.

File: c10/cuda/CUDACachingAllocator.cpp
Comment (o-hanna) [reply c10/cuda/CUDACachingAllocator.cpp]: Done. Thank you.

File: c10/cuda/CUDACachingAllocator.cpp
Comment (o-hanna) [reply c10/cuda/CUDACachingAllocator.cpp]: I tested this and there is no big difference in the running time. I let the vector handle resizing now. Thank you.

**Comment by Jack47** [comment]
Hi @o-hanna, I'm very impressive on this PR because I'm facing fragmentation problems too. I believe  visualizing the allocate/free event sequences will greatly help our dig why fragmentation happens during training iterations. I want to try this pr and what to know do you have some good visualizing tools to display those sequences?

ps: I'm implementing DTR on 1.11, find it's performance is hugely impacted by memory fragmentation.

File: c10/cuda/CUDACachingAllocator.cpp
Comment (mrshenli) [inline c10/cuda/CUDACachingAllocator.cpp:495]: nit: if we only care about allocate/free, we can save a type arg by converting size into a sized integer. 

This is also what profiler does: https://github.com/pytorch/pytorch/blob/6d23ef847e1c119939095ca1e8ce195e697b2a09/c10/cuda/CUDACachingAllocator.cpp#L794

File: c10/cuda/CUDACachingAllocator.cpp
Comment (mrshenli) [inline c10/cuda/CUDACachingAllocator.cpp:498]: since we already have device in the event struct, do we still need to make `alloc_free_events` use device as the 1st layer of index?

File: c10/cuda/CUDACachingAllocator.cpp
Comment (mrshenli) [inline c10/cuda/CUDACachingAllocator.cpp:506]: I saw you removed mutex here. You can keep it by making it as mutable.

File: c10/cuda/CUDACachingAllocator.cpp
Comment (mrshenli) [inline c10/cuda/CUDACachingAllocator.cpp:1649]: since this is after allocating the block, it won't capture the last block that caused the OOM. To capture that, you can  use a invalid ptr as the initial value and update it after the block is created. See D38065728

File: torch/cuda/__init__.py
Comment (mrshenli) [inline torch/cuda/__init__.py:378]: there are 2 new links between other functions

File: torch/cuda/__init__.py
Comment (mrshenli) [inline torch/cuda/__init__.py:366]: write some test for this?

**Comment by o-hanna** [comment]
> Hi @o-hanna, I'm very impressive on this PR because I'm facing fragmentation problems too. I believe visualizing the allocate/free event sequences will greatly help us to dig out why fragmentation happens during training iterations. I want to try this pr and what to know do you have some good visualizing tools to display those sequences?
> 
> ps: I'm implementing DTR on 1.11, find it's performance is hugely impacted by memory fragmentation.

Hi @Jack47, thank you for your comment. We are using matplotlib to visualize the the events. For example the function "plot_memory_events" will plot the events, where "AllocFreeEvents" is the returned sequence of events, "device" is the index of device, "alloc_free_seq_lengths" is a vector containing number of events for each batch to make a separation between the events for each batch (this is optional), "num_events" is also optional if you do not need to plot all events, "distance" caps large empty memory spaces in the plot to value distance to give more space for the events.
[memory_events_4batches.pdf](https://github.com/pytorch/pytorch/files/9223142/memory_events_4batches.pdf)


An example of the plot is attached to this comment:
- The red lines separate the events across different batches.
- The sequence number (on Y-axis) is increased whenever an allocate event is followed by free or a free event is followed by allocate. In other words a run of allocate events (or a run of free events) are assigned to the same sequence no.
- Different events are assigned random colors.

Please note that some events might appear to overlap on the edges in the plot. This is caused by the matplotlib adding a frame to the rectangles.

Please let me know if you have any questions/comments :)


def plot_memory_events(AllocFreeEvents, device, alloc_free_seq_lengths = [], num_events=0, distance=int(1e6)):
    events_for_device = AllocFreeEvents[device]
    if num_events <=0:
        num_events = len(events_for_device)

    events = [[i,events_for_device[i]] for i in range(num_events)] # attaching sequence number to do sort then reverse sort
    
    sorted_events = sorted(events,key=lambda x: x[1].ptr) # sort by pointer

    # subtract min pointer value to start from 0

    next_ptr_dist = sorted_events[0][1].ptr
    if next_ptr_dist > distance:
        for j in range(len(sorted_events)):
            sorted_events[j][1].ptr -= next_ptr_dist

    # caping large white spaces at "distance" and compute sequence length
    T = 1 # length of sequence
    for i in range(len(sorted_events)-1):
        next_ptr_dist = sorted_events[i+1][1].ptr - (sorted_events[i][1].ptr+sorted_events[i][1].size)
        if next_ptr_dist > distance:
            for j in range(i+1,len(sorted_events)):
                sorted_events[j][1].ptr += distance-next_ptr_dist
        if events[i][1].type != events[i+1][1].type:
            T += 1

    # reverse the sort
    for x in sorted_events:
        events[x[0]] = x[1]

    min_address = 0
    max_address = max(x[1].ptr+x[1].size for x in sorted_events)
    
    # plot events
    _, ax = plt.subplots()
    
    t = 0
    num_events_to_plot = num_events # plot last num_events_to_plot events
    status = events[num_events-num_events_to_plot].type # 1 for alloc
    batch_idx = 0
    for j in range(num_events-num_events_to_plot, num_events):
        if events[j].type != status:
            status = events[j].type
            t = t+1
        if status == 0:
            col = "white"
        else:
            col = (random.uniform(0, 1), random.uniform(0, 1), random.uniform(0, 1))
        
        ax.add_patch(Rectangle((events[j].ptr, t), events[j].size-1, T-t,color=col))
        col = 'k'
        linwidth = 0.5
        if (len(alloc_free_seq_lengths)>0) and (j==alloc_free_seq_lengths[batch_idx]):
            col = 'red'
            linwidth = 1
            batch_idx += 1
        plt.axhline (y = t, xmin =min_address, xmax = max_address, color=col, linewidth = linwidth)
    plt.xlim([min_address,max_address])
    plt.ylim([0,t+1])
    plt.xlabel("Memory status")
    plt.ylabel("Sequence no.")
    plt.title("Memory events")
    plt.show()
    plt.savefig('memory_events.pdf')

File: c10/cuda/CUDACachingAllocator.cpp
Comment (o-hanna) [reply c10/cuda/CUDACachingAllocator.cpp:1649]: I had this in a previous commit but I removed it because I though when an OOM occurs, the program will stop and the allocate/free sequence vector will be destroyed? But I am not sure. I was thinking that alternatively in M2 we could save the allocate/free sequence in a file and the API reads from that file? Please let me know what you think. Thank you.

File: c10/cuda/CUDACachingAllocator.cpp
Comment (o-hanna) [reply c10/cuda/CUDACachingAllocator.cpp:495]: Done. Thank you.

File: c10/cuda/CUDACachingAllocator.cpp
Comment (o-hanna) [reply c10/cuda/CUDACachingAllocator.cpp:494]: I removed the device from the event struct. I think it is more convenient to separate events for different devices? Please let me know if you prefer to use one layer for the list of events and include the device in the event struct. Thank you.

File: c10/cuda/CUDACachingAllocator.cpp
Comment (o-hanna) [reply c10/cuda/CUDACachingAllocator.cpp:501]: Done. Thank you.

File: torch/cuda/__init__.py
Comment (o-hanna) [reply torch/cuda/__init__.py:366]: Done. Thank you.

File: torch/cuda/__init__.py
Comment (o-hanna) [reply torch/cuda/__init__.py:378]: Done. Thank you.

**Comment by o-hanna** [comment]
Update: the member variable size of event struct takes negative values to indicate a free event. The type and device member variables are removed from the event struct. The PR description is updated to include the changes.

**Comment by Jack47** [comment]
> 

cool，great thanks for sharing plot code!

**Comment by HamidShojanazeri** [comment]
Thanks @o-hanna, this is great, a very much needed tool for debugging. I was wondering if we could also expose some more info to the graph that show what kind of event has triggered the mem allocation, e.g all_gather etc. This would be helpful to better identify the bottlenecks in the training job. cc:@mrshenli

File: c10/cuda/CUDACachingAllocator.cpp
Comment (mrshenli) [reply c10/cuda/CUDACachingAllocator.cpp:1642]: > I had this in a previous commit but I removed it because I though when an OOM occurs, the program will stop and the allocate/free sequence vector will be destroyed?

No, the Python side will through an OOM `RuntimeError`, and we can capture that to avoid process crash. As long as the process is still alive, the per-process singleton `CUDACachingAllocator` and its states should be still available .

File: c10/cuda/CUDACachingAllocator.cpp
Comment (mrshenli) [reply c10/cuda/CUDACachingAllocator.cpp:494]: I don't have strong opinion on this one. Please feel free to use the one you prefer.

File: c10/cuda/CUDACachingAllocator.cpp
Comment (mrshenli) [inline c10/cuda/CUDACachingAllocator.cpp:1080]: some additional thoughts. IIUC, currently, we are recording allocte/free events at the allocator level instead of cudaMalloc/cudaFree level, i.e., we gonna record an allocate event even if the request is served by a cached block. If this is the case, in followup PRs, shall we add a mode to capture raw cudaMalloc/cudaFree events. This might be more helpful to debug OOM problems.

File: c10/cuda/CUDACachingAllocator.cpp
Comment (o-hanna) [reply c10/cuda/CUDACachingAllocator.cpp:494]: Thank you.

File: c10/cuda/CUDACachingAllocator.cpp
Comment (o-hanna) [reply c10/cuda/CUDACachingAllocator.cpp:1080]: That is a great idea. I can also add labels to allocate events, e.g., served by by a cached block, served after freeing a cached block, served after defragmentation etc?
I added this on the to-do list. Thank you

File: test/test_cuda.py
Comment (o-hanna) [reply test/test_cuda.py:296]: I added it in case some workers don't fully clean up, so this could free some memory before the allocation? I can remove it if you think it is not necessary

File: test/test_cuda.py
Comment (o-hanna) [reply test/test_cuda.py]: I was thinking to get output from both test allocate and test free even if the allocate test fails by returning negative size. I saw this in other tests, so I thought it is preferred to get output from multiple asserts at the same time; but it is not important. I removed it.

File: c10/cuda/CUDACachingAllocator.cpp
Comment (o-hanna) [reply c10/cuda/CUDACachingAllocator.cpp:1643]: Done. Thank you

File: .gitignore
Comment (o-hanna) [reply .gitignore]: Done. Thank you.

File: test/test_cuda.py
Comment (o-hanna) [reply test/test_cuda.py]: If the allocation fails? (It is initialized with None) I know it is unlikely to happen here :) I removed it since we removed the try-finally

File: test/test_cuda.py
Comment (o-hanna) [inline test/test_cuda.py:308]: I removed this but the push is taking some time.

File: test/test_cuda.py
Comment (mrshenli) [inline test/test_cuda.py:304]: any reason for testing >0 instead of ==1? When will we get a different size?

Comment (o-hanna) [reply test/test_cuda.py:304]: The returned size is actually rounded up to 512. I can test for this.

Comment (o-hanna) [reply test/test_cuda.py:304]: Done. Thank you. It was because of previous implementation returned the actual size of allocated block which can be different.

**Comment by soumith** [comment]
we should have @zdevito take a look as well before merging

**Comment by mrshenli** [comment]
Just to clarify, this PR is not targeting master branch. This is mainly to unblock @o-hanna to explore caching algorithms in his own branch.

**Comment by lw** [comment]
I imagine that the hook framework that @sypneiwski is working on should allow to do this tracing fully in Python, without having to specialize the CUDA caching allocator for this specific purpose.

@sypneiwski do you have a PR to share for them to take a look?

**Comment by albanD** [comment]
The PR that @lw was refering to: https://github.com/pytorch/pytorch/pull/82824

**Pull request merged**
